doc/python_src_layout_for_aws_lambdas.md

```diff
@@ -0,0 +1,337 @@
+---
+title: Python src Layout for AWS Lambdas
+menu_order: 1
+post_status: draft
+tags: aws, awscommunity, python, poetry, awslambda
+post_excerpt: How to create AWS Lambdas with Poetry and the Python src layout
+canonical_url: https://informed.iq/python_src_layout_for_aws_lambdas
+---
+
+# Python src Layout for AWS Lambdas
+
+Over the last several years, the Python community has been moving towards a packaging layout format known as the `src layout` as the recommended way to organize directories and files for Python Packages.
+
+The main takeaways of the `src layout` are:
+
+* Your project has a `src` folder
+  * The `src` folder only has sub folder[s]. Each sub folder
+    * Is a Python Package
+    * Its name is the name of the package
+    * It contains an `__init__.py` (that's pretty much what makes it a Python Package)
+  * All Python source code that makes up your appplication (other than tests or build utilities) for the project are in these sub folder[s]
+* Modern Python tooling expects / works best with this layout
+* Makes packaging easier and less prone to errors
+* "editable" installs for development and regular installation for testing
+* Generally no need for "import trickery"
+* References at end of this article
+
+We were looking for standards and best practices for organizing our Python AWS Lambda projects using [Poetry](https://python-poetry.org) in our monorepo and kept coming across the `src layout` as the recommended way to organize Python projects that had multiple modules. At the same time just about all examples of Python Lambda projects did not use `src layout` and in fact seemed to want to have the handler module at the top level of the project directory.  There just weren't many examples of this combination of `src layout` `Poetry` and AWS Lambdas.
+
+Turns  out that it wasn't really difficult. The main "trick" is setting the Lambda `handler` name correctly.
+
+> All this example code is available at  [Informed/blogpost-python-src-layout](https://github.com/Informed/blogpost-python-src-layout)
+ 
+## Example src  layout for an AWS Lambda
+
+Here's an example `src layout` suitable for AWS Lambda
+
+* This is a section of what could be a regular or monorepo that shows one lambda project: `my_lambda`
+  * You could have more lambda projects under services, each would be a separate subproject with the same layout
+    * How to implement such a multi-project monorepo is left to a future blog post
```

Comment:
Definitely interested in seeing this monorepo blog post. Curious if you are deploying each individual Lambda Function separately or as a larger backend service made up multiple Lambda Functions along with additional infrastructure.

Reply:
I'm definitely interested in doing the blog post. I'm going to try to get it done this week, but we'll see if that comes true or not :-) 

We deploy each lambda service individually. They each have their own set of CircleCI workflows. We have a matching set of CI  jobs and CLI jobs to deploy the services via Terraform and to deploy the code via direct AWS CLI/SDK commands.

I recently was interviewed and gave an overview of the CI/CD aspects on [SHOWOFF CI/CD - CircleCI pipeline for Serverless platform with Terraform and Rob J. Berger](https://youtu.be/_YYrwOR4Xzc)

Reply:
Thx for the info.